src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,132 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    internal partial class DocumentOutlineViewModel : INotifyPropertyChanged, IDisposable
+    {
+        private readonly ILanguageServiceBroker2 _languageServiceBroker;
+        private readonly VisualStudioCodeWindowInfoService _visualStudioCodeWindowInfoService;
+        private readonly CompilationAvailableTaggerEventSource _textViewEventSource;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private CancellationToken CancellationToken => _cancellationTokenSource.Token;
+
+        public DocumentOutlineViewModel(
+            ILanguageServiceBroker2 languageServiceBroker,
+            IAsynchronousOperationListener asyncListener,
+            VisualStudioCodeWindowInfoService visualStudioCodeWindowInfoService,
+            CompilationAvailableTaggerEventSource textViewEventSource)
+        {
+            _languageServiceBroker = languageServiceBroker;
+            _visualStudioCodeWindowInfoService = visualStudioCodeWindowInfoService;
+            _textViewEventSource = textViewEventSource;
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            // initialize public properties
+            _sortOption = SortOption.Location;
+            _documentSymbolUIItems = new ObservableCollection<DocumentSymbolUIItem>();
+
+            // setup work queues
+            _documentSymbolQueue = new AsyncBatchingWorkQueue<VisualStudioCodeWindowInfo, DocumentSymbolDataModel?>(
+                DelayTimeSpan.Short,
+                GetDocumentSymbolAsync,
+                EqualityComparer<VisualStudioCodeWindowInfo>.Default,
+                asyncListener,
+                CancellationToken);
+
+            _filterAndSortQueue = new AsyncBatchingWorkQueue<FilterAndSortOptions, DocumentSymbolDataModel?>(
+                DelayTimeSpan.NearImmediate,
+                FilterAndSortDataModelAsync,
+                EqualityComparer<FilterAndSortOptions>.Default,
+                asyncListener,
+                CancellationToken);
+
+            _updateUIQueue = new AsyncBatchingWorkQueue<UIData>(
+                DelayTimeSpan.NearImmediate,
+                UpdateUIAsync,
+                asyncListener,
+                CancellationToken);
+
+            _textViewEventSource.Changed += OnEventSourceChanged;
+            _textViewEventSource.Connect();
+
+            // queue initial model update
+            EnqueueModelUpdateAsync().Forget();
```

Comment:
surprised that we have an queue method that returns a task.  Is there a reason we need to have it do that?